File: next-app/src/lib/algorithms/adaptive-coach.ts
/**
 * Adaptive Coach Service — port of services/adaptive_coach.py.
 * Implements MacroFactor-style weekly check-ins with auto-adjusting targets.
 */

import { calculateWeightTrend } from "./weight-trend";

const KCAL_PER_KG = 7700.0;
const MAX_WEEKLY_KCAL_CHANGE = 200;
const MIN_WEIGHIN_DAYS = 3;
const MIN_DIARY_DAYS = 4;
const MIN_KCAL_TARGET = 1200;
const MAX_KCAL_TARGET = 6000;

export type WeightGoal = "lose" | "maintain" | "gain";

export type CheckInStatus = "ready" | "insufficient_data" | "not_enough_time" | "error";

export interface MacroPreset {
  key: string;
  displayName: string;
  description: string;
  proteinPct: number;
  carbsPct: number;
  fatPct: number;
}

export const PRESET_MAP: Record<string, MacroPreset> = {
  low_carb: {
    key: "low_carb",
    displayName: "Low Carb",
    description: "Bajo en carbohidratos, alto en grasas",
    proteinPct: 0.25,
    carbsPct: 0.45,
    fatPct: 0.30,
  },
  balanced: {
    key: "balanced",
    displayName: "Balanceado",
    description: "Distribucion equilibrada",
    proteinPct: 0.30,
    carbsPct: 0.35,
    fatPct: 0.35,
  },
  high_protein: {
    key: "high_protein",
    displayName: "High Protein",
    description: "Alto en proteinas para ganancia muscular",
    proteinPct: 0.40,
    carbsPct: 0.30,
    fatPct: 0.30,
  },
  high_carb: {
    key: "high_carb",
    displayName: "High Carb",
    description: "Alto en carbohidratos para energia",
    proteinPct: 0.25,
    carbsPct: 0.50,
    fatPct: 0.25,
  },
  keto: {
    key: "keto",
    displayName: "Keto",
    description: "Muy bajo en carbohidratos",
    proteinPct: 0.30,
    carbsPct: 0.05,
    fatPct: 0.65,
  },
  custom: {
    key: "custom",
    displayName: "Personalizado",
    description: "Macros ajustados manualmente",
    proteinPct: 0.30,
    carbsPct: 0.35,
    fatPct: 0.35,
  },
};

export interface MacroGrams {
  protein: number;
  carbs: number;
  fat: number;
  totalKcal: number;
}

export function calculateMacroGrams(preset: MacroPreset, totalKcal: number): MacroGrams {
  const protein = Math.round((totalKcal * preset.proteinPct) / 4);
  const carbs = Math.round((totalKcal * preset.carbsPct) / 4);
  const fat = Math.round((totalKcal * preset.fatPct) / 9);
  return { protein, carbs, fat, totalKcal: protein * 4 + carbs * 4 + fat * 9 };
}

export interface WeeklyData {
  startDate: string;
  endDate: string;
  avgDailyKcal: number;
  trendWeightStart: number;
  trendWeightEnd: number;
  daysWithDiary: number;
  daysWithWeighins: number;
  trendChangeKg: number;
  calculatedTdee: number;
  hasEnoughData: boolean;
}

export interface CheckInExplanation {
  line1: string;
  line2: string;
  line3: string;
  line4: string;
  line5: string;
  allLines: string[];
}

export interface CheckInResult {
  status: CheckInStatus;
  weeklyData: WeeklyData;
  estimatedTdee: number;
  proposedKcalTarget: number;
  proposedMacros: MacroGrams;
  explanation: CheckInExplanation;
  wasClamped: boolean;
  dailyAdjustmentKcal: number;
  errorMessage?: string;
}

export function getDailyAdjustment(goal: WeightGoal, weeklyRateKg: number): number {
  const adjustment = Math.round((weeklyRateKg * KCAL_PER_KG) / 7);
  if (goal === "lose") return -Math.abs(adjustment);
  if (goal === "gain") return Math.abs(adjustment);
  return 0;
}

export function calculateCheckin(
  plan: {
    goal: WeightGoal;
    weekly_rate_kg: number;
    current_kcal_target: number | null;
    initial_tdee: number;
    macro_preset: string;
  },
  weeklyData: WeeklyData
): CheckInResult {
  const emptyMacros: MacroGrams = { protein: 0, carbs: 0, fat: 0, totalKcal: 0 };

  if (!weeklyData.hasEnoughData) {
    return {
      status: "insufficient_data",
      weeklyData,
      estimatedTdee: 0,
      proposedKcalTarget: 2000,
      proposedMacros: emptyMacros,
      explanation: {
        line1: "",
        line2: "",
        line3: "",
        line4: "",
        line5: "",
        allLines: [],
      },
      wasClamped: false,
      dailyAdjustmentKcal: 0,
      errorMessage: `Se necesitan al menos ${MIN_DIARY_DAYS} dias de diario y ${MIN_WEIGHIN_DAYS} pesajes. Tienes: ${weeklyData.daysWithDiary} dias de diario, ${weeklyData.daysWithWeighins} pesajes.`,
    };
  }

  const calculatedTdee = weeklyData.calculatedTdee;
  const adjustment = getDailyAdjustment(plan.goal, plan.weekly_rate_kg);
  let newTarget = Math.round(calculatedTdee + adjustment);

  // Clamp: max weekly change
  let wasClamped = false;
  const previousTarget = plan.current_kcal_target || plan.initial_tdee;
  const minAllowed = previousTarget - MAX_WEEKLY_KCAL_CHANGE;
  const maxAllowed = previousTarget + MAX_WEEKLY_KCAL_CHANGE;

  if (newTarget < minAllowed) {
    newTarget = minAllowed;
    wasClamped = true;
  } else if (newTarget > maxAllowed) {
    newTarget = maxAllowed;
    wasClamped = true;
  }

  if (newTarget < MIN_KCAL_TARGET) {
    newTarget = MIN_KCAL_TARGET;
    wasClamped = true;
  } else if (newTarget > MAX_KCAL_TARGET) {
    newTarget = MAX_KCAL_TARGET;
    wasClamped = true;
  }

  // Calculate macros
  const preset = PRESET_MAP[plan.macro_preset] || PRESET_MAP.balanced;
  const macroGrams = calculateMacroGrams(preset, newTarget);

  // Build explanation
  const sign = weeklyData.trendChangeKg > 0 ? "+" : "";
  const adjText =
    adjustment < 0
      ? `${adjustment}kcal (deficit)`
      : adjustment > 0
        ? `+${adjustment}kcal (superavit)`
        : "0kcal (mantenimiento)";

  const explanation: CheckInExplanation = {
    line1: `Ingesta media: ${Math.round(weeklyData.avgDailyKcal)} kcal/dia`,
    line2: `Cambio de trend: ${sign}${weeklyData.trendChangeKg.toFixed(2)} kg`,
    line3: `TDEE estimado: ${Math.round(calculatedTdee)} kcal`,
    line4: `Ajuste objetivo: ${adjText}`,
    line5: `Nuevo target: ${newTarget} kcal`,
    allLines: [],
  };
  explanation.allLines = [
    explanation.line1,
    explanation.line2,
    explanation.line3,
    explanation.line4,
    explanation.line5,
  ];

  return {
    status: "ready",
    weeklyData,
    estimatedTdee: Math.round(calculatedTdee),
    proposedKcalTarget: newTarget,
    proposedMacros: macroGrams,
    explanation,
    wasClamped,
    dailyAdjustmentKcal: adjustment,
  };
}

export function buildWeeklyData(
  weightEntries: { fecha: string; peso: number | null }[],
  foodDailyTotals: { fecha: string; kcal: number }[],
  startDate: string,
  endDate: string
): WeeklyData {
  // Filter to period
  const periodWeights = weightEntries.filter(
    (w) => w.peso != null && w.fecha >= startDate && w.fecha <= endDate
  );
  const periodFood = foodDailyTotals.filter(
    (f) => f.fecha >= startDate && f.fecha <= endDate
  );

  // Avg daily kcal
  const kcalValues = periodFood.filter((f) => f.kcal > 0).map((f) => f.kcal);
  const avgKcal =
    kcalValues.length > 0
      ? kcalValues.reduce((a, b) => a + b, 0) / kcalValues.length
      : 0;

  // Trend weights
  let trendStart = 0;
  let trendEnd = 0;

  const allWeights = weightEntries.filter((w) => w.peso != null);
  if (allWeights.length >= 2) {
    const result = calculateWeightTrend(allWeights);
    trendEnd = result.emaWeight;

    const earlyWeights = allWeights.filter((w) => w.fecha <= startDate);
    if (earlyWeights.length >= 2) {
      const earlyResult = calculateWeightTrend(earlyWeights);
      trendStart = earlyResult.emaWeight;
    } else if (earlyWeights.length > 0) {
      trendStart = earlyWeights[earlyWeights.length - 1].peso!;
    } else {
      trendStart = trendEnd;
    }
  } else if (allWeights.length > 0) {
    trendStart = trendEnd = allWeights[0].peso!;
  }

  const trendChangeKg = trendEnd - trendStart;
  const startD = new Date(startDate + "T00:00:00");
  const endD = new Date(endDate + "T00:00:00");
  const days = Math.max(1, (endD.getTime() - startD.getTime()) / (1000 * 60 * 60 * 24));
  const calculatedTdee = avgKcal - (trendChangeKg * KCAL_PER_KG) / days;

  return {
    startDate,
    endDate,
    avgDailyKcal: avgKcal,
    trendWeightStart: trendStart,
    trendWeightEnd: trendEnd,
    daysWithDiary: kcalValues.length,
    daysWithWeighins: periodWeights.length,
    trendChangeKg,
    calculatedTdee,
    hasEnoughData:
      kcalValues.length >= MIN_DIARY_DAYS && periodWeights.length >= MIN_WEIGHIN_DAYS,
  };
}

File: services/adaptive_coach.py
"""
Adaptive Coach Service — port of Juan-Tracker's adaptive_coach_service.dart.
Implements MacroFactor-style weekly check-ins with auto-adjusting targets.
"""

from dataclasses import dataclass, field
from enum import Enum
from typing import Optional
from datetime import date, timedelta

KCAL_PER_KG = 7700.0
MAX_WEEKLY_KCAL_CHANGE = 200
MIN_WEIGHIN_DAYS = 3
MIN_DIARY_DAYS = 4
MIN_KCAL_TARGET = 1200
MAX_KCAL_TARGET = 6000


class WeightGoal(Enum):
    LOSE = "lose"
    MAINTAIN = "maintain"
    GAIN = "gain"


class CheckInStatus(Enum):
    READY = "ready"
    INSUFFICIENT_DATA = "insufficient_data"
    NOT_ENOUGH_TIME = "not_enough_time"
    ERROR = "error"


class MacroPreset(Enum):
    LOW_CARB = ("Low Carb", "Bajo en carbohidratos, alto en grasas", 0.25, 0.45, 0.30)
    BALANCED = ("Balanceado", "Distribucion equilibrada", 0.30, 0.35, 0.35)
    HIGH_PROTEIN = ("High Protein", "Alto en proteinas para ganancia muscular", 0.40, 0.30, 0.30)
    HIGH_CARB = ("High Carb", "Alto en carbohidratos para energia", 0.25, 0.50, 0.25)
    KETO = ("Keto", "Muy bajo en carbohidratos", 0.30, 0.05, 0.65)
    CUSTOM = ("Personalizado", "Macros ajustados manualmente", 0.30, 0.35, 0.35)

    def __init__(self, display_name, description, protein_pct, carbs_pct, fat_pct):
        self.display_name = display_name
        self.description = description
        self.protein_pct = protein_pct
        self.carbs_pct = carbs_pct
        self.fat_pct = fat_pct

    def calculate_grams(self, total_kcal):
        """Calculate macro grams from total kcal."""
        protein_g = round((total_kcal * self.protein_pct) / 4)
        carbs_g = round((total_kcal * self.carbs_pct) / 4)
        fat_g = round((total_kcal * self.fat_pct) / 9)
        return MacroGrams(protein=protein_g, carbs=carbs_g, fat=fat_g)


PRESET_MAP = {
    "low_carb": MacroPreset.LOW_CARB,
    "balanced": MacroPreset.BALANCED,
    "high_protein": MacroPreset.HIGH_PROTEIN,
    "high_carb": MacroPreset.HIGH_CARB,
    "keto": MacroPreset.KETO,
    "custom": MacroPreset.CUSTOM,
}


@dataclass
class MacroGrams:
    protein: int
    carbs: int
    fat: int

    @property
    def total_kcal(self):
        return (self.protein * 4) + (self.carbs * 4) + (self.fat * 9)


@dataclass
class WeeklyData:
    start_date: date
    end_date: date
    avg_daily_kcal: float
    trend_weight_start: float
    trend_weight_end: float
    days_with_diary: int
    days_with_weighins: int

    @property
    def trend_change_kg(self):
        return self.trend_weight_end - self.trend_weight_start

    @property
    def calculated_tdee(self):
        """TDEE = avg_kcal - (delta_trend_weight * 7700 / days)"""
        days = (self.end_date - self.start_date).days
        if days <= 0:
            return self.avg_daily_kcal
        delta_kcal = self.trend_change_kg * KCAL_PER_KG / days
        return self.avg_daily_kcal - delta_kcal

    @property
    def has_enough_data(self):
        return self.days_with_diary >= MIN_DIARY_DAYS and self.days_with_weighins >= MIN_WEIGHIN_DAYS


@dataclass
class CheckInExplanation:
    line1: str = ""  # Ingesta media
    line2: str = ""  # Cambio de peso
    line3: str = ""  # TDEE estimado
    line4: str = ""  # Ajuste objetivo
    line5: str = ""  # Nuevo target

    @property
    def all_lines(self):
        return [self.line1, self.line2, self.line3, self.line4, self.line5]


@dataclass
class CheckInResult:
    status: CheckInStatus
    weekly_data: WeeklyData
    estimated_tdee: int = 0
    proposed_kcal_target: int = 2000
    proposed_macros: MacroGrams = field(default_factory=lambda: MacroGrams(0, 0, 0))
    explanation: CheckInExplanation = field(default_factory=CheckInExplanation)
    was_clamped: bool = False
    daily_adjustment_kcal: int = 0
    error_message: Optional[str] = None


def get_daily_adjustment(goal, weekly_rate_kg):
    """Calculate daily kcal adjustment from goal and weekly rate."""
    # kg/week * 7700 kcal/kg / 7 days = kcal/day
    adjustment = round(weekly_rate_kg * KCAL_PER_KG / 7)
    if goal == WeightGoal.LOSE or goal == "lose":
        return -abs(adjustment)
    elif goal == WeightGoal.GAIN or goal == "gain":
        return abs(adjustment)
    return 0


def calculate_checkin(plan, weekly_data):
    """
    Calculate weekly check-in.
    plan: dict from db (coach_plans row)
    weekly_data: WeeklyData instance
    Returns CheckInResult.
    """
    # Validate data
    if not weekly_data.has_enough_data:
        return CheckInResult(
            status=CheckInStatus.INSUFFICIENT_DATA,
            weekly_data=weekly_data,
            error_message=(
                f"Se necesitan al menos {MIN_DIARY_DAYS} dias de diario "
                f"y {MIN_WEIGHIN_DAYS} pesajes. "
                f"Tienes: {weekly_data.days_with_diary} dias de diario, "
                f"{weekly_data.days_with_weighins} pesajes."
            ),
        )

    # Calculate TDEE from data
    calculated_tdee = weekly_data.calculated_tdee

    # Get goal adjustment
    goal = plan["goal"]
    weekly_rate = plan["weekly_rate_kg"]
    adjustment = get_daily_adjustment(goal, weekly_rate)

    # New target = TDEE + adjustment (negative for deficit)
    new_target = round(calculated_tdee + adjustment)

    # Clamp: max weekly change
    was_clamped = False
    previous_target = plan.get("current_kcal_target") or plan["initial_tdee"]

    min_allowed = previous_target - MAX_WEEKLY_KCAL_CHANGE
    max_allowed = previous_target + MAX_WEEKLY_KCAL_CHANGE

    if new_target < min_allowed:
        new_target = min_allowed
        was_clamped = True
    elif new_target > max_allowed:
        new_target = max_allowed
        was_clamped = True

    # Clamp: absolute health limits
    if new_target < MIN_KCAL_TARGET:
        new_target = MIN_KCAL_TARGET
        was_clamped = True
    elif new_target > MAX_KCAL_TARGET:
        new_target = MAX_KCAL_TARGET
        was_clamped = True

    # Calculate macros
    preset_name = plan.get("macro_preset", "balanced")
    preset = PRESET_MAP.get(preset_name, MacroPreset.BALANCED)
    macro_grams = preset.calculate_grams(new_target)

    # Build explanation
    sign = "+" if weekly_data.trend_change_kg > 0 else ""
    adj_text = f"{adjustment}kcal (deficit)" if adjustment < 0 else (
        f"+{adjustment}kcal (superavit)" if adjustment > 0 else "0kcal (mantenimiento)"
    )

    explanation = CheckInExplanation(
        line1=f"Ingesta media: {round(weekly_data.avg_daily_kcal)} kcal/dia",
        line2=f"Cambio de trend: {sign}{weekly_data.trend_change_kg:.2f} kg",
        line3=f"TDEE estimado: {round(calculated_tdee)} kcal",
        line4=f"Ajuste objetivo: {adj_text}",
        line5=f"Nuevo target: {new_target} kcal",
    )

    return CheckInResult(
        status=CheckInStatus.READY,
        weekly_data=weekly_data,
        estimated_tdee=round(calculated_tdee),
        proposed_kcal_target=new_target,
        proposed_macros=macro_grams,
        explanation=explanation,
        was_clamped=was_clamped,
        daily_adjustment_kcal=adjustment,
    )


def build_weekly_data(weight_entries, food_daily_totals, start_date, end_date, trend_calculator=None):
    """
    Build WeeklyData from raw entries.
    weight_entries: list of dicts with fecha, peso
    food_daily_totals: list of dicts with fecha, kcal
    """
    from services.weight_trend import WeightTrendCalculator

    if trend_calculator is None:
        trend_calculator = WeightTrendCalculator()

    # Filter to period
    period_weights = [
        w for w in weight_entries
        if w.get("peso") is not None and start_date <= _parse_date(w["fecha"]) <= end_date
    ]
    period_food = [
        f for f in food_daily_totals
        if start_date <= _parse_date(f["fecha"]) <= end_date
    ]

    # Avg daily kcal
    kcal_values = [f["kcal"] for f in period_food if f.get("kcal") and f["kcal"] > 0]
    avg_kcal = sum(kcal_values) / len(kcal_values) if kcal_values else 0

    # Trend weights
    trend_start = 0
    trend_end = 0

    all_weights = [w for w in weight_entries if w.get("peso") is not None]
    if len(all_weights) >= 2:
        result = trend_calculator.calculate(all_weights)
        trend_end = result.ema_weight

        # For start, use EMA up to start_date
        early_weights = [
            w for w in all_weights if _parse_date(w["fecha"]) <= start_date
        ]
        if len(early_weights) >= 2:
            early_result = trend_calculator.calculate(early_weights)
            trend_start = early_result.ema_weight
        elif early_weights:
            trend_start = early_weights[-1]["peso"]
        else:
            trend_start = trend_end
    elif all_weights:
        trend_start = trend_end = all_weights[0]["peso"]

    return WeeklyData(
        start_date=start_date,
        end_date=end_date,
        avg_daily_kcal=avg_kcal,
        trend_weight_start=trend_start,
        trend_weight_end=trend_end,
        days_with_diary=len(kcal_values),
        days_with_weighins=len(period_weights),
    )


def _parse_date(d):
    """Parse date from string or date object."""
    if isinstance(d, date):
        return d
    from datetime import datetime
    try:
        return datetime.fromisoformat(str(d)).date()
    except (ValueError, TypeError):
        return date.today()
